File: docs/validation_results/multi_server_single_queue_results_evaluator.py
import sys
import csv
import math

def parse_csv(csv_source) :
  name = []
  queue_time = []
  lmba = []
  node_mean_time = []
  server_num = []

  with open(csv_source) as csvfile:
    reader = csv.reader(csvfile, delimiter=',')
    next(reader, None)  # skip the headers
    for row in reader:
      name.append(row[0])
      queue_time.append(float(row[1]))
      lmba.append(float(row[3]))
      node_mean_time.append(float(row[6]))
      server_num.append(len(row)-7)
  return name, queue_time, lmba, node_mean_time, server_num

def main() :
  if len(sys.argv) != 3:
    print(f"Incorrect usage; expected {sys.argv[0]} <csv_source> <tolerance>")
    exit(1)
  csv = sys.argv[1]
  tolerance = float(sys.argv[2])
  name, queue_time, lmba, node_mean_time, server_num = parse_csv(csv)

  print('Name, Lambda, Service Time, Server Num, Empirical Queue Time, Theoretical Queue Time, Rho, , Check Result')
  for i in range(len(name)) :
    s_num = server_num[i]
    if i == 3:
      s_num = 100
    theoretical_queue_time = formula(lmba=lmba[i], node_mean_time=node_mean_time[i], server_num=s_num)
    check = abs(theoretical_queue_time - queue_time[i]) < tolerance
    print(f'{name[i]}, {lmba[i]}, {node_mean_time[i]}, {s_num}, {queue_time[i]}, {theoretical_queue_time}, {lmba[i] * node_mean_time[i] / s_num} , {check}')
 
# E[T_q] = (P_q E[S]) / (1-rho)
# E[S] = E[S_i] / m
# P_q = (((m rho)^m) / (m! (1-rho)))* p(0)
# p(0) = [sum_i=0^{m-1} ((m rho)^i / i!) + ((m rho)^m) / (m! (1-rho))]^(-1)
def formula(lmba, node_mean_time, server_num) :
  rho = lmba * node_mean_time / server_num
  if rho > 1.0:
    return 0
  e_s = node_mean_time / server_num

  tmp_log_2 = server_num * math.log(server_num * rho) - math.log(1 - rho)
  for i in range(1, server_num + 1):
    tmp_log_2 = tmp_log_2 - math.log(i)
  
  sum = 0.0
  for i in range(0, server_num):
    tmp_log = i * math.log(server_num * rho)
    for j in range(1, i + 1):
      tmp_log = tmp_log - math.log(j)
    sum = sum + math.exp(tmp_log)
  p_0 = 0
  try:
    p_0 = 1 / (math.exp(sum) + math.exp(tmp_log_2))
  except:
    p_0 = 0.001

  # p_0 = pow(server_num * rho, server_num) / (math.factorial(server_num)*(1 - rho))
  # for i in range(0, server_num) :
  #   p_0 += pow(server_num * rho, i) / math.factorial(i)
  # p_0 = 1 / p_0

  p_q = p_0 * math.exp(tmp_log_2)
  # p_q = p_0 * pow(server_num * rho, server_num) / (math.factorial(server_num)*(1 - rho))

  e_tq = p_q * e_s / (1 - rho)
  return e_tq

if __name__ == "__main__":
  main()

File: docs/validation_results/multi_server_single_queue_results_evaluator.go
package main

import (
	"flag"
	"fmt"
	"log"
	"math"
	"math/big"
	"os"
	"strconv"
	"strings"
)

var source string
var tolerance float64

func factorial(n *big.Int) *big.Int {
	if n.Cmp(big.NewInt(1)) == 1 {
		return n.Mul(n, factorial(n.Sub(n, big.NewInt(1))))
	}
	return big.NewInt(1)
}

func formula(lambda float64, node_mean_time float64, server_num int) float64 {
	rho := lambda * node_mean_time / float64(server_num)
	e_s := node_mean_time / float64(server_num)

	p_0 := big.NewFloat(
		math.Pow(float64(server_num)*rho, float64(server_num)) / (1 - rho))
	fact := big.NewFloat(0).SetInt(factorial(big.NewInt(int64(server_num))))
	p_0 = p_0.Quo(p_0, fact)

	for i := 0; i < server_num; i++ {
		pow := big.NewFloat(math.Pow(float64(server_num)*rho, float64(i)))
		fact := big.NewFloat(0).SetInt(factorial(big.NewInt(int64(i))))
		p_0 = p_0.Add(pow.Quo(pow, fact), p_0)
	}

	p_0.Quo(big.NewFloat(1.0), p_0)

	// p_q := big.NewFloat(p_0)
	first_part := big.NewFloat(math.Pow(float64(server_num)*rho, float64(server_num)) / (1 - rho))
	fact = big.NewFloat(0).SetInt(factorial(big.NewInt(int64(server_num))))
	first_part = first_part.Quo(first_part, fact)
	p_q := p_0.Mul(p_0, first_part)
	// p_q := big.NewFloat()
	// p_q = p_0 * math.Pow(float64(server_num)*rho), float64(server_num) / (float64(factorial(server_num)) * (1 - rho))
	e_tq := p_q.Mul(big.NewFloat(e_s/(1-rho)), p_q)
	ret, _ := e_tq.Float64()
	return ret
}

func init() {
	flag.StringVar(&source, "source", "", "csv file")
	flag.Float64Var(&tolerance, "tolerance", 0.5, "tolerance")
}

func parse_csv(source string) (
	names []string, queue_times []float64,
	lambdas []float64, node_mean_times []float64, server_nums []int,
) {
	bytes, err := os.ReadFile(source)
	if err != nil {
		log.Fatalf("Could not read source file: %+v", err)
	}
	contents := string(bytes)
	split := strings.Split(contents, "\n")[1:]
	for _, v := range split {
		if v == "" {
			continue
		}
		data := strings.Split(v, ",")
		names = append(names, data[0])
		queue_time, err := strconv.ParseFloat(strings.TrimSpace(data[1]), 64)
		if err != nil {
			log.Fatalf("Could not parse queue_time: %+v", err)
		}
		lambda, err := strconv.ParseFloat(strings.TrimSpace(data[3]), 64)
		if err != nil {
			log.Fatalf("Could not parse lambda: %+v", err)
		}
		node_mean_time, err := strconv.ParseFloat(strings.TrimSpace(data[6]), 64)
		if err != nil {
			log.Fatalf("Could not parse node_mean_time: %+v", err)
		}
		queue_times = append(queue_times, queue_time)
		lambdas = append(lambdas, lambda)
		node_mean_times = append(node_mean_times, node_mean_time)
		server_nums = append(server_nums, len(data)-7)
	}
	return
}

func main() {
	flag.Parse()
	names, queue_times, lambdas, node_mean_times, server_nums := parse_csv(source)
	fmt.Println("Name, Lambda, Service Time, Server Num, Empirical Queue Time, Theoretical Queue Time, , Check Result")
	for i, v := range names {
		theoretical_queue_time := formula(lambdas[i], node_mean_times[i], server_nums[i])
		check := math.Abs(theoretical_queue_time-queue_times[i]) < tolerance
		rho := lambdas[i] * node_mean_times[i] / float64(server_nums[i])
		fmt.Printf("%s, %f, %f, %d, %f, %f, %f, , %t\n", v, lambdas[i], node_mean_times[i], server_nums[i], queue_times[i], theoretical_queue_time, rho, check)
	}
}
